lib/cylc/tests/test_param_expand.py

```diff
@@ -0,0 +1,209 @@
+             ('foo_i0_j2', {'i': 0, 'j': 2}),
+             ('foo_i1_j0', {'i': 1, 'j': 0}),
+             ('foo_i1_j1', {'i': 1, 'j': 1}),
+             ('foo_i1_j2', {'i': 1, 'j': 2})]
+        )
+
+    def test_name_two_names(self):
+        """Test name expansion for two names."""
+        self.assertEqual(
+            self.name_expander.expand('foo<i>, bar<j>'),
+            [('foo_i0', {'i': 0}),
+             ('foo_i1', {'i': 1}),
+             ('bar_j0', {'j': 0}),
+             ('bar_j1', {'j': 1}),
+             ('bar_j2', {'j': 2})]
+        )
+
+    def test_name_specific_val_1(self):
+        """Test singling out a specific value, in name expansion."""
+        self.assertEqual(
+            self.name_expander.expand('foo<i=0>'),
+            [('foo_i0', {'i': 0})]
+        )
+
+    def test_name_specific_val_2(self):
+        """Test specific value in the first parameter of a pair."""
+        self.assertEqual(
+            self.name_expander.expand('foo<i=0,j>'),
+            [('foo_i0_j0', {'i': 0, 'j': 0}),
+             ('foo_i0_j1', {'i': 0, 'j': 1}),
+             ('foo_i0_j2', {'i': 0, 'j': 2})]
+        )
+
+    def test_name_specific_val_3(self):
+        """Test specific value in the second parameter of a pair."""
+        self.assertEqual(
+            self.name_expander.expand('foo<i,j=1>'),
+            [('foo_i0_j1', {'i': 0, 'j': 1}),
+             ('foo_i1_j1', {'i': 1, 'j': 1})]
+        )
+
+    def test_name_fail_bare_value(self):
+        """Test foo<0,j> fails."""
+        # It should be foo<i=0,j>.
+        self.assertRaises(ParamExpandError,
+                          self.name_expander.expand, 'foo<0,j>')
+
+    def test_name_fail_undefined_param(self):
+        """Test that an undefined parameter gets failed."""
+        # m is not defined.
+        self.assertRaises(ParamExpandError,
+                          self.name_expander.expand, 'foo<m,j>')
+
+    def test_name_fail_param_value_too_high(self):
+        """Test that an out-of-range parameter gets failed."""
+        # i stops at 3.
+        self.assertRaises(ParamExpandError,
+                          self.name_expander.expand, 'foo<i=4,j>')
+
+    def test_name_multiple(self):
+        """Test expansion of two names, with one and two parameters."""
+        self.assertEqual(
+            self.name_expander.expand('foo<i>, bar<i,j>'),
+            [('foo_i0', {'i': 0}),
+             ('foo_i1', {'i': 1}),
+             ('bar_i0_j0', {'i': 0, 'j': 0}),
+             ('bar_i0_j1', {'i': 0, 'j': 1}),
+             ('bar_i0_j2', {'i': 0, 'j': 2}),
+             ('bar_i1_j0', {'i': 1, 'j': 0}),
+             ('bar_i1_j1', {'i': 1, 'j': 1}),
+             ('bar_i1_j2', {'i': 1, 'j': 2})]
+        )
+
+    def test_graph_expand_1(self):
+        """Test graph expansion with two parameters each side of an arrow."""
+        self.assertEqual(
+            self.graph_expander.expand("bar<i,j>=>baz<i,j>"),
+            {"bar_i0_j1=>baz_i0_j1",
```

Comment:
Sadly we can't get away with Python 3 syntax just yet! See https://github.com/kinow/cylc/pull/1.

Reply:
Oh, sorry, I tested only on 2.7! Will have to remember the paragraph from our user guide 

>Python 2 >= 2.6 is required. Python 2 >= 2.7.9 is recommended for the best security. Python 2 should already be installed in your Linux system. https://python.org/. 

Merging it and updating this pull request.